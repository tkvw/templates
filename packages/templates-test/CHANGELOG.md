# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.3](https://github.com/tkvw/templates/compare/v1.0.2...v1.0.3) (2019-10-03)

**Note:** Version bump only for package @tkvw/templates-test





## [1.0.2](https://github.com/tkvw/templates/compare/v1.0.1...v1.0.2) (2019-10-03)

**Note:** Version bump only for package @tkvw/templates-test





## 1.0.1 (2019-10-03)

**Note:** Version bump only for package @tkvw/templates-test
