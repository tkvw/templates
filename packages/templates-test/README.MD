# Template-test

Tool to create snapshots from template to showcase template usage. 
