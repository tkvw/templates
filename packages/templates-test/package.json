{
  "name": "@tkvw/templates-test",
  "version": "1.0.1",
  "description": "Test utility for testing templates",
  "main": "dist/main/index.js",
  "scripts": {
    "build": "run-s clean && run-p build:*",
    "build:main": "tsc -p ./tsconfig.build.json",
    "clean": "trash dist"
  },
  "keywords": [],
  "author": "Dennie de Lange",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tkvw/templates.git",
    "directory": "packages/templates-test"
  },
  "license": "ISC",
  "dependencies": {
    "@tkvw/templates": "file:../templates",
    "@types/jest": "^24.0.18",
    "jest": "^24.9.0"
  }
}
