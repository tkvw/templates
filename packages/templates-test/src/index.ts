import * as path from 'path';
import * as fse from 'fs-extra';
import { runner } from '@tkvw/templates';

export const testTemplates = async (
  templates: any,
  {
    cwd,
    paths = [],
    ...options
  }: {
    cwd: string;
    paths?: string[];
  },
) => {
  if (Array.isArray(templates)) {
    for (const test of templates) {
      if (Array.isArray(test)) {
        const [description, ...args] = test;
        it(description, async () => {
          const testDir = path.resolve(cwd, '../generated', paths.join('-'));
          await fse.ensureDir(testDir);
          const generatedDir = path.resolve(testDir, description);
          await fse.emptyDir(generatedDir);
          const processedInstructions = await runner([...paths, ...args], {
            destination: generatedDir,
            quiet: true,
          });
          const renderedTemplateInstructions = processedInstructions.filter(pi => pi.actions.find(a => a.template));
          for (const renderedTemplateInstruction of renderedTemplateInstructions) {
            if (typeof renderedTemplateInstruction.destination === 'string') {
              const renderedTemplate = await fse.readFile(renderedTemplateInstruction.destination, {
                encoding: 'utf8',
              });
              expect(renderedTemplate).toMatchSnapshot(description);
            }
          }
          expect(true).toBeTruthy();
        });
      }
    }
  } else if (typeof templates === 'object') {
    const keys = Object.keys(templates);
    for (const key of keys) {
      const value = templates[key];

      describe(key, () => {
        testTemplates(value, {
          ...options,
          cwd,
          paths: [...paths, key],
        });
      });
    }
  }
};
