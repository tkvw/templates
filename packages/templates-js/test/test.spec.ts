import { testTemplates } from '@tkvw/templates-test';

describe('',()=>{
  testTemplates(
    {
      js: {
        gatsby: {
          plugin: [['with name only', '--name', 'dennie']],
        },
        monorepo: [['with name only', '--name', 'dennie']],
        test: {
          jest: {
            unit: [['simple test', '--files', 'simple.ts,another.ts']],
          },
        },
      },
    },
    {
      cwd: __dirname,
    },
  );
});
