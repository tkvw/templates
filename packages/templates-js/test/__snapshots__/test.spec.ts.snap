// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` js gatsby plugin with name only: with name only 1`] = `
"{
  \\"name\\": \\"dennie\\",
  \\"publishConfig\\": {
    \\"access\\": \\"public\\"
  },
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"// TODO add a plugin description\\",
  \\"homepage\\": \\"// TODO https://github.com/tkvw/dennie\\",
  \\"main\\": \\"index.js\\",
  \\"scripts\\": {
    \\"clean\\": \\"rm -rf ./dist\\",
    \\"build\\": \\"npm run clean && tsc -p ./tsconfig.build.json && cp ./package.json ./dist\\",
    \\"deploy\\": \\"npm run build && cd dist && npm publish\\"
  },
  \\"keywords\\": [\\"gatsby\\", \\"gatsby-plugin\\", \\"// TODO: Please enter additional keywords here\\"],
  \\"author\\": \\"Dennie de Lange\\",
  \\"license\\": \\"// TODO: Enter license type\\",
  \\"peerDependencies\\": {
    \\"gatsby\\": \\"*\\"
  },
  \\"devDependencies\\": {
    \\"gatsby\\": \\"*\\"
  }
}
"
`;

exports[` js monorepo with name only: with name only 1`] = `
"{
  \\"private\\": true,
  \\"name\\": \\"dennie\\",
  \\"scripts\\": {
    \\"clean\\": \\"lerna clean\\",
    \\"bootstrap\\": \\"lerna bootstrap\\",
    \\"test\\": \\"CI=true lerna run test --stream\\",
    \\"start\\": \\"lerna run --parallel start\\",
    \\"prepare\\": \\"lerna run prepare\\",
    \\"build\\": \\"npm run prepare && lerna run --parallel build\\"
  },
  \\"devDependencies\\": {
    \\"lerna\\": \\"*\\"
  },
  \\"workspaces\\": [\\"packages/*\\"],
  \\"lint-staged\\": {
    \\"*.{ts,tsx,js,jsx}\\": [\\"eslint --fix\\", \\"git add\\"]
  },
  \\"husky\\": {
    \\"hooks\\": {
      \\"pre-commit\\": \\"lint-staged\\"
    }
  }
}
"
`;

exports[` js test jest unit simple test: simple test 1`] = `
"import another from \\"./another\\";

describe('another', () => {
  it(\\"should succeed\\", async () => {
    // TODO: Add test
  })
})
"
`;
